### PR TITLE
Add additional properties to WC Analytics events

### LINKED ISSUE #7301: Add new events to the Analytics module
**Description:** we want to take advantage of the existing Event Analytics module (p90Yrv-RL-p2) to improve our understanding of the shopper behavior at checkout and help merchants to improve their checkout conv. rate. We need to create a new event and update the woocommerceanalytics_product_purchase event to include more event props. 
Ventures made some updates in this tracking last year: pb0Spc-ou-p2

**Acceptance criteria:**

- [ ] we need to add a new event to track the checkout view and all the payment options available:

`woocommerceanalytics_product_checkout_view`
`payment_options`: an array with the options available. Eg [paypal, stripe, cod, amazon]
`express_checkou`t: an array with the options available. Eg [apple_pay] or [null] if they don’t display them
`guest_checkout`: Yes/No. Identify if the store allows users to checkout without an account
`create_account`: Yes/No Identify if the store allows users to create an account on checkout
`device`: mobile/desktop

- [ ] We can use the `woocommerceanalytics_product_purchase` event to track more information when the user finishes the purchase. We need to add the following event props:
`payment_option`: the option that was used. Eg Stripe or COD
`express_checkout`: if the express checkout option was used: Eg: Apple Pay or Null
`guest_checkout`: if the guest checkout was used
`create_account`: check if the shopper created an account (Y/N)
`device`: mobile/desktop

### PR DESCRIPTION
**Pull request by moon0326**
~Please do not merge this PR yet. Changes in this PR depend on the newly added action in WC Pay. We need to merge these changes after releasing the WC Pay version 2.9.~ WC Pay version 2.9 has been [released](https://github.com/Automattic/woocommerce-payments/releases/tag/2.9.0) and I've updated this PR to work with previous versions of WC Payments in 4e9a4fb071145687942c8bdb6a552c6026e483ff

Fixes https://github.com/woocommerce/woocommerce-admin/issues/7301

## Changes proposed in this Pull Request:

- Supports `array` value to be added as an event property https://github.com/Automattic/jetpack/commit/4f2d1bce8ded004302e0bd78a4340e233df3c440 - This is needed to add `payment_options` value.
-  Added additional properties to `woocommerceanalytics_product_purchase` and `woocommerceanalytics_product_checkout_view`. The additional properties are listed in the [issue](https://github.com/woocommerce/woocommerce-admin/issues/7301).
- One notable change is that it uses an [action](https://github.com/Automattic/woocommerce-payments/pull/2610).
 from WC Pay when the express checkout methods are enabled.

#### Does this pull request change what data or activity we track or use?

We're adding the following properties to `woocommerceanalytics_product_purchase` and `woocommerceanalytics_product_checkout_view` events.

#### woocommerceanalytics_product_purchase

**payment_option**: the option that was used. Eg Stripe or COD
**express_checkout**: if the express checkout option was used: Eg: Apple Pay or Null
**guest_checkout**: if the guest checkout was used
**create_account**: check if the shopper created an account (Y/N)
**device**: mobile/desktop

#### woocommerceanalytics_product_checkout_view

**payment_options**: an array with the options available. Eg [paypal, stripe, cod, amazon]
**express_checkout**: an array with the options available. Eg [apple_pay] or [null] if they don’t display them
**guest_checkout**: Yes/No. Identify if the store allows users to checkout without an account
**create_account**: Yes/No Identify if the store allows users to create an account on checkout
**device**: mobile/desktop

## Testing instructions:

Please use [WC Pay 2.9.0](https://github.com/Automattic/woocommerce-payments/releases/tag/2.9.0)

If you're new to WC Analytic testing, please perform the following steps to enable WC Analytic.

1. Install and activate WooCommerce.
2. Finish onboarding wizard and enable `Direct bank transfer` in WooCommerce -> Settings -> Payments.
2. Install and activate Jetpack.
3. Connect your site to Jetpack
4. WC Analytic doesn't track in test/dev mode. Open this [file](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/modules/woocommerce-analytics/class-jetpack-woocommerce-analytics.php#L41) and add `return true` to force WC Analytic to track during the testing.
5. Click on a product on `/shop` to see the details.
6. Open your browser inspector -> Network -> search for "t.gif" then refresh.
7. You should see `t.gif` request.

If you don't see `t.gif` request for some reason, make sure WooCommerce Analytics module is activated in `/wp-admin/admin.php?page=jetpack_modules`

**Case 1** -- Without WC Payment installed/activated

1. Make sure WC Payment plugin is **not** installed/activated.
2. Add a product to the cart.
3. Open browser inspector -> Network and search for `t.gif`
4. Continue to the checkout page.
5. You should see `t.gif` request.
6. Click on the request and confirm that the request URL contains all the required properties for `woocommerceanalytics_product_checkout_view` event.
7. Click [ Place order ] button with `Direct bank transfer` selected.
8. You should see another `t.gif` request on the `Order received` page for `woocommerceanalytics_product_purchase` event. Confirm that the required properties are included in the URL.

Bonus: Add multiple products and confirm you get `t.gif` request for each product.

Screen Shot 2021-08-25 at 11 46 03 AMth_

You can use tools like https://www.freeformatter.com/url-parser-query-string-splitter.html to parse the URL into a table.

**Case 2** -- WC Payment activated + express checkout methods enabled.

1. Use ngrok to get `https` and publicly accessible address.
2. Install, activate, and connect WC Payment
3. Make sure `Express Checkouts` are enabled in Payments -> Settings
4. Repeat steps 2~8 from Case 1.
5. Confirm both events have `express_checkout` value.

`woocommerceanalytics_product_checkout_view` should have either `google_pay` or `apple_pay`
`woocommerceanalytics_product_purchase` should have `null`

**Case 3** -- WC Payment activated and express checkout methods disabled.

1. Use ngrok to get `https` and publicly accessible address.
2. Install, activate, and connect WC Payment
3. Navigate to Payments -> Settings and disable `Express Checkouts`
4. Repeat steps 2~8 from Case 1
5. Confirm `express_checkout` value is `null` in both events.

**Case 4** -- WC Payment activated + express checkout methods enabled + checkout using Google/Apple pay.

1. Use ngrok to get `https` and publicly accessible address.
2. Install, activate, and connect WC Payment
3. Navigate to Payments -> Settings and disable `Express Checkouts`
4. Create a new product and set the price to $0.5 (minimum price to use google pay).
5. Add the product to the cart and contitnue to the checkout page.
6. Click Google Pay or Apple Pay to checkout. 
7. Confirm `express_checkout` value.

both events should have `google_pay` or `apple_pay`

**Case 5** -- WC Payment version lower than 2.9.0.

The action used in this PR was added in WC Payment version 2.9.0. If the user has WC Payment version lower than 2.9.0, then the action should not be used.

1. Install & activated WC Payment version lower than 2.9.0
2. Add a product to the cart and continue to the checkout page.
3. On the checkout page, right-click on the page and click "View Page Source"
4. Scroll to the bottom of the page and confirm the event JS script is not using `wp.hooks.addAction`

### CODE REVIEW

**Review by jeherve — COMMENTED** [review]
> Please do not merge this PR yet. Changes in this PR depend on the newly added action in WC Pay. We need to merge these changes after releasing the WC Pay version 2.9.
> 
> we would love to include this PR in the next release on September 7th

When do you think this new version of WCPay will be released? Can we merge this PR before next Monday, so it can be included in the next release?

**Review by haszari — COMMENTED** [review]
Overall this looks reasonable. I've added a few questions about the details. Note, I'm not clear on what questions we want to answer with this, so would like to hear more about that.

Before going further I'd like to confirm what questions we need to answer with this data, and check in with the data team (cc @georgekng @deltaWhiskey) about how best to do that. We track related information in WooCommerce Tracker (e.g. gateways), so may not need to add to tracks events. There's a lot of generic info added to these new events, some may potentially be redundant.

Another thing that I'm not clear about – will these new events work consistently with cart & checkout blocks vs shortcodes? At a minimum we should test to confirm if they work and if the data is all there.

Related, if we want to track which express payment options are presented to a customer, then we might need to consider other plugins that add express payments (like Stripe).

**Review by adrianduffell — DISMISSED** [review]
This tested great for me, thank you @moon0326 !

Approving this from a Woo Start + Manage perspective 🚀 

I didn't get to test Case 4 as it needed a real credit card set up with Apple Pay. Maybe a sandbox account should be used for this? https://developer.apple.com/apple-pay/sandbox-testing/

The only possible issue I found was in case 3, on the checkout screen I got an empty value instead of `null`. i.e `express_checkout=&...`. I was using Safari for my testing.

**Comment by haszari** [comment]
> WC Analytic doesn't track in test/dev mode. Open this file and add return true to force WC Analytic to track during the testing.

@moon0326 FYI you can also test without hacking the code – just use a different browser session to purchase as a guest or non-admin user.

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (haszari) [reply projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php]: I think we should make them as clear as possible to start – will propose tweaks in my review :)

Also I think it's worth getting a review from data team (more important than my review really). Ideally it would be good to have a P2 outlining exactly what questions we need to answer, so we can ensure we are tracking the right info, and nothing redundant.

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (haszari) [reply projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:338]: Fair enough – though let's try and improve this in the next iteration. Hopefully we can avoid release pressure next time.

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (haszari) [reply projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:294]: I'm curious if Woo has an API for express payment methods. Ideally we could call a standard API to determine what express options are available, and avoid tight coupling to specific gateway plugins.

WooCommerce Blocks checkout block has some support/APIs for express payment options, so might be worth discussing with that team for advice.

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (haszari) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:373]: Let's standardise the format of boolean props (`1 0`, `Y N`, `yes no` etc). 

Note that elsewhere in this file, `1 0` is used for boolean props (e.g. `cart_page_contains_cart_block`). This is compact and unambiguous; I'd recommend sticking with this, or whatever data team recommends.

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (haszari) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:378]: This block of code is tightly coupled to WooCommerce payments. Perhaps it could be factored out into a function so it's clear it depends on another plugin. 

Or perhaps use a hook/filter to provide a standard interface to any plugin that might need to supply this info (e.g. other gateway plugins).

Also see above comment – if there's a Woo API for express payment methods, we should use that. If there isn't an API - should we add one?

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (haszari) [reply projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:397]: Best to check in with data team about how we plan to segment data / analyse by device (mobile, screen size, etc) before adding a new prop. Tracks events may already have a generic mechanism for this; `User Agent` is sent with all tracks requests.

File: projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php
Comment (haszari) [inline projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:625]: Is this the best way to detect that a new account was requested? Maybe there's another hook fired when the account is created.

**Comment by moon0326** [comment]
> > WC Analytic doesn't track in test/dev mode. Open this file and add return true to force WC Analytic to track during the testing.
> 
> @moon0326 FYI you can also test without hacking the code – just use a different browser session to purchase as a guest or non-admin user.

Thank you! Looking at the code, that should work as well.

Could you take a look at the comments I left to your comments? 🙇  I want to ship the current changes tomorrow (code freeze) then follow up with Stripe plugin and JS code inclusion changes in the next release.

**Review by haszari — DISMISSED** [review]
I gave this a really quick test – there are lots of combinations to test, and they are all somewhat complex to set up!

Here's what I tested:

- Checkout event shows correct store options (create account, allow guest, available payment options).
- When an express option available, it's included in checkout event. Google Pay is the only one I can get going: `express_checkout: google_pay`. 
- Purchase with wcpay shows correct gateway `woocommerce_payments`.
- Only tested with core payment options (COD and Cheque), and WooCommerce Payments.
- WooCommerce Payments was in test mode.

I didn't test the following:

- If these events and props all work correctly with checkout or cart blocks from WooCommerce Blocks.
- Create account on purchase flow.
- Any express payments – Google pay fails for me (`Billing Street address is a required field. Billing Town / City is a required field. Billing Postcode is a required field.`).

### Feedback on events and props
There's a lot of new info getting tracked here. That's great if we need it all - though I'm curious if some of this info might be redundant or not essential. Store settings are also tracked in WooCommerce Tracker. 

After we ship new events and props it's harder to remove things or adjust naming, especially if the intention and use is not clearly defined.

Let's follow this up – @pmcpinto could you P2 the question(s) you want to answer with this info>

Regarding the actual naming of the new props, here are some suggestions. The goal here is to make each prop as clear and unambiguous as possible.

- `woocommerceanalytics_product_checkout`
   - ~~`payment_options`~~ `available_payment_options` or `available_regular_payment_options` (to distinguish from express)
   - ~~`express_checkout `~~ `available_express_payment_options`
   - ~~`guest_checkout`~~ `store_allows_guest_checkout`
   - ~~`create_account`~~ `store_allows_create_account_at_checkout`
   - ~~`device`~~ tbc - let's clarify what we are tracking here; if we stick with `wp_is_mobile` perhaps we can name `is_mobile` or `useragent_is_mobile`
- `woocommerceanalytics_product_purchase`
   - ~~`payment_option`~~ `user_payment_option` or `user_regular_payment_option` if this is empty/null for express
   - ~~`express_checkout`~~ `user_express_payment_option` 
   - Or - would `user_payment_option` and `is_user_payment_option_express` work better?
   - ~~`guest_checkout`~~ `is_guest_checkout`
   - ~~`create_account`~~ `is_checkout_create_account` or similar
   - otherwise use same props as `checkout` event

For all these properties, please use standard types and formats as recommended by data team (e.g. boolean props).

### Blocks and other express checkout 
Noting a couple of places that I haven't tested:

- Express payment on product page.
- Express payment on cart page.
- Checkout page using block.

If we want to track checkout/purchase anywhere they happen on a site, we should test those flows too (and follow up if needed).

### Next steps
I've approved this PR – if it's urgent to get this shipped in next Jetpack, so we can start run funnels, I don't want to hold it up.

At the same time, I'd love to find out more about what questions this tracking will answer so we can ensure we're collecting the right info (and no more). If it's not urgent, maybe it's better to have that discussion before shipping. cc @pmcpinto

**Review by jeherve — DISMISSED** [review]
this looks good to me. 🚢

**Comment by moon0326** [comment]
> this looks good to me. 🚢

Thank you @jeherve 👍 I'll give one last test shortly and merge it before the end of today.

**Review by sdixon194 — APPROVED** [review]

**Pull request merged**
